File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    private int[] memo = new int[101];
    private int solve(int i, int j, int[] nums) {
        if (i >= j) {
            return 0;
        }
        if (memo[i] != -1) {
            return memo[i];
        }
        int take = nums[i] + solve(i+2, j, nums);
        int skip = solve(i+1, j, nums);
        return memo[i] = Math.max(take, skip);
    }
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 1) {
            return nums[0];
        }
        if (n == 2) {
            return Math.max(nums[0], nums[1]);
        }
        Arrays.fill(memo, -1);
        int takeZero = solve(0, n-1, nums);
        Arrays.fill(memo, -1);
        int takeOne  = solve(1, n, nums);
        return Math.max(takeZero, takeOne);
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        if n == 2:
            return max(nums[0], nums[1])
        
        def solve(nums):
            memo = [-1] * len(nums)
            def dp(i):
                if i >= len(nums):
                    return 0
                if memo[i] != -1:
                    return memo[i]
                take = nums[i] + dp(i + 2)
                skip = dp(i + 1)
                memo[i] = max(take, skip)
                return memo[i]
            return dp(0)
        
        zero = solve(nums[:-1])
        one = solve(nums[1:])
        
        return max(zero, one)